### ISSUE TITLE
Exploring Simple Siamese Representation Learning

### ISSUE THREAD

**Issue by pocca2048** [pocca2048/ML-paper-reading#4]
**Conference** : 
**Link** : https://arxiv.org/pdf/2011.10566.pdf
**Authors' Affiliation** : Facebook AI Research, Kaiming He
**TL;DR** : 사실 negative pair, large batch, momentum encoder 중 어느 것도 없이도 충분히 의미있는 representation을 학습할 수 있다. Stop-gradient operation이 collapsing을 방지하는데 핵심적인 역할을 한다.

**Summary** :
# 1. Introduction

Siamese network = weights sharing NN applied on 2+ inputs

하나의 이미지에 대한 2개의 augmentation을 인풋으로 하고 similarity를 최대화.

그러나 샴네트워크는 하나의 constant로 모든 output이 collapse해버릴수 있다.

그동안 collapsing을 막는 여러 방법들이 제안되었다: contrastive learning에서는 pos는 끌어당기고  neg는 밀어서 negative pair를 통해 constant solution을 배제한다. clustering을 사용하는 방법도 있다( SwAV). BYOL에서는 positive pair에만 의존하지만 momentum encoder를 사용해서 collapse를 방지한다.

이 논문에서는 위의 어떤 방법도 사용하지 않고 간단한 샴 네트워크만으로 collapsing을 방지할 수 있음을 보인다.

> We empirically show that collapsing solutions *do* exist, but a **stop-gradient** operation (Figure 1) is critical to pre- vent such solutions.

stop-gradient가 중요하다는 것은 기저에 사실 다른 optimization problem이 풀어지고 있다는걸 암시한다.

가설: 2개의 변수 집합이 implicit하게 있고, SimSiam은 각 집합을 번갈아가며 optimize 하는것처럼 행동한다.

이 가설을 검증하기 위해 PoC 실험을 한다.

샴 네트워크는 자연스럽게 invariance를 모델링 하기위한 inductive bias를 제시한다.

convolution이 자연스러운 translation-invariance를 도입하는 것처럼 샴네트워크는 더 복잡한 transformation에 대한 invaraicne를 제공한다.

# 2. Related Work

# 3. Method

```python
# f: backbone + projection mlp
# h: prediction mlp
for x in loader: # load a minibatch x with n samples
   x1, x2 = aug(x), aug(x) # random augmentation
   z1, z2 = f(x1), f(x2) # projections, n-by-d
   p1, p2 = h(z1), h(z2) # predictions, n-by-d
   L = D(p1, z2)/2 + D(p2, z1)/2 # loss
   L.backward() # back-propagate
   update(f, h) # SGD update
def D(p, z): # negative cosine similarity
   z = z.detach() # stop gradient
   p = normalize(p, dim=1) # l2-normalize
   z = normalize(z, dim=1) # l2-normalize
   return -(p*z).sum(dim=1).mean()
```

backbone + projection + prediction

Prediction MLP head의 역할은 한 view의 output을 다른 view에게 맞도록 transform해주는 것.

negative cosine similarity를 minimize하는 것은 l2-normalized vector의 MSE를 minimize 하는 것과 같다 (up to a scale of 2?)

# 4. Empirical Study

## 4.1 Stop-gradient

없으면 무조건 collapse함. variance가 0이되고 있으면 unit sphere에 랜덤하게 분포.

## 4.2 Predictor

h를 없애면 작동안함. 즉, h는 identity mapping 임(?????). 사실 이런 관측은 symmetric loss가 사용될때 예상된거임 (?)

h를 제거한 loss를 사용하면 collapsing 이 관측됨.

asymmetric loss를 사용해도 h를 없애면 fail한다고 함. 왜지????

h만 lr decay를 안쓰는게 더 효과적. h는 항상 latest representation에 adapt해야 하므로 lr 을 줄여서 수렴하도록 강요하는게 필요치 않다.

## 4.3 Batch Size

대부분의 bsz에서 잘 작동. 64, 128 이어도 비교적 잘 작동하고 이보다 크면 비슷.

근데 신기하게도 SGD optimizer는 supervised 에서도 그렇고 batch가 너무 크면 잘 작동 안함. 성능이 4096일때 급격히 떨어짐.

## 4.4 Batch Normalization

bn 아예 안쓰면 acc이 절반으로. projection에 까지 쓰면 최대. prediction에도 쓰면 unstable. 그냥 다른 supervised때와 똑같음.

BN이 collapsing을 막아준다는 증거는 전혀 없음.

## 4.5 Similarity Function

cosine similarity 대신에 $D(p_1, z_2) = -softmax(z_2) \log softmax(p_1)$ 으로 두면 acc 5%정도 떨어짐.

collapsing은 cosine similarity 자체의 문제가 아님.

## 4.6 Symmetrization

asymmetric 버전 쓰면 4%정도 떨어짐. 하지만 collapsing 이슈랑은 관계없음.

symmetrization은 각 이미지에 대해 하나 많은 prediction을 하므로 이걸 각 이미지마다 sampling을 2pair 만큼 해서 보충하면 acc gap이 줄어든다.

## 4.7 Summary

stop gradient가 collapsing을 멈추는 가장 중요한 요소이다.

# 5 Hypothesis

SimSiam에 의해 최적화되는게 무엇인가에 대한 가설을 세우고 PoC 실험을 한다.

## 5.1 Formulation

가설은 simsiam이 EM 같은 알고리즘의 implementation이라는 것이다. EM은 2개의 변수 집합을 가지고 2개의 기저하는 subproblem을 푼다.

stop-gradient의 존재는 extra set of variable을 introduce하는 것의 결과다.
